### ISSUE TITLE
Rendering videos onto FBOs needs their ID

### ISSUE THREAD

**Issue by Boscop** [glium/glium#1613]
I want to render videos onto FBOs to use in an opengl scene (e.g. drawing video onto the faces of a cube).
I'm using libmpv which [draws on a given FBO, but it needs its ID](https://github.com/Cobrand/mpv-rs/blob/master/src/mpv_handler.rs#L199).
So I want to create a SimpleFramebuffer and render the video onto that, but how can I get its ID to pass to `mpv.draw()`?

**Comment by tomaka** [comment]
The trait `GlObject` is meant to provide you with IDs for objects.
However since framebuffer objects can be invalidated at any time if the user calls `rebuild_context`, the trait can't be implemented properly.

Also note that it is probably a bad idea to let another library manipulate your current OpenGL context, unless that library saves and restores the state of the context, which is almost never the case with these libraries.

**Comment by Boscop** [comment]
If I'm never calling `rebuild_context`, is there a hacky way to still get the ID for `SimpleFramebuffer`?
And is there a way to save the state of the current context with glium before I let mpv draw, and then restore it afterwards?

**Comment by tomaka** [comment]
> is there a hacky way to still get the ID for SimpleFramebuffer?

No, but it's something that could be added.

> And is there a way to save the state of the current context with glium before I let mpv draw, and then restore it afterwards?

It's the contrary that must happen. Mpv must save the state of the current context, then restore it.
What could be added to glium as well is some sort of `invalidate_gl_state()` method, but I'm not totally sure it's possible to implement.

**Comment by Boscop** [comment]
Yes, it would be very useful to be able to get the ID of the framebuffer.

In the meantime, can I guess its ID, in other words, are framebuffer IDs generated deterministically?
E.g. if the main screen has ID 0, does the first framebuffer get ID 1, then 2 etc.?

The way I understand the context issue is that what should happen is:
```js
my app code...
{ // semantic scope for everything that mpv does
    let tmp_state = gl_state;
    mpv modifies gl_state
    gl_state = tmp_state;
}
my app code...
```
But then why wouldn't it work like this?
```js
my app code...
let tmp_state = gl_state;
{ // semantic scope for everything that mpv does
    mpv modifies gl_state
}
gl_state = tmp_state;
my app code...
```
In other words, why wouldn't it work if I save and restore the GL state around mpv's modifying it?

**Comment by tomaka** [comment]
> In the meantime, can I guess its ID, in other words, are framebuffer IDs generated deterministically?

Not at all.

> In other words, why wouldn't it work if I save and restore the GL state around mpv's modifying it?

I wasn't seeing it the same way as you.
What I was thinking with "save and restore the state" is to keep track of what we modified and restore it afterwards.

Presumably what mpv does is simply write to a texture, so it just needs `glBindTexture` and `glActiveTexture` I guess. Maybe it uses a buffer, in which case there's also a `glBindBuffer`. In both cases these are easy to restore afterwards.

On the other side the state that glium keeps track of [is very large](https://github.com/glium/glium/blob/master/src/context/state.rs#L8-L331). If you were to ask glium to call OpenGL function in order to "restore" its state, then you'd need dozens of function calls.

**Comment by HookedBehemoth** [comment]
libmpv makes promises about the state it restores and modifies:
```c
/** OpenGL state
 * ------------
 *
 * OpenGL has a large amount of implicit state. All the mpv functions mentioned
 * above expect that the OpenGL state is reasonably set to OpenGL standard
 * defaults. Likewise, mpv will attempt to leave the OpenGL context with
 * standard defaults. The following state is excluded from this:
 *
 *      - the glViewport state
 *      - the glScissor state (but GL_SCISSOR_TEST is in its default value)
 *      - glBlendFuncSeparate() state (but GL_BLEND is in its default value)
 *      - glClearColor() state
 *      - mpv may overwrite the callback set with glDebugMessageCallback()
 *      - mpv always disables GL_DITHER at init
 *
 * Messing with the state could be avoided by creating shared OpenGL contexts,
 * but this is avoided for the sake of compatibility and interoperability.
**/
 ```
<https://github.com/mpv-player/mpv/blob/master/libmpv/render_gl.h#L46-L62>
